chore: adr for store enhancement (#933)
defenseunicorns/pepr: adr/00080-store-key-sanitization.md

> # 8: Store Key Sanitization

Reviewer: ```suggestion
# 8: Store Key Encoding
```
The word "sanitization" usually describes swapping out / cleaning up of something & leaving only the "good stuff" behind, but... this feature isn't really changing what's input so the word "sanitization" seems kinda out of place to me.

For example, [here](https://www.esecurityplanet.com/endpoint/prevent-web-attacks-using-input-sanitization/#:~:text=Sanitizing%20consists%20of%20removing%20any,before%20insertion%20in%20a%20database.): 
"Sanitizing consists of removing any unsafe characters from user inputs, and validating will check to see if the data is in the expected format and type. Sanitizing modifies the input to ensure it's in a valid format for display, or before insertion in a database."

Seems to me that the word "encoding" better describes exactly what we're doing AND also implies some stuff of the downstream consequences (e.g. that the Store will no longer be readable as-is / will need to be decoded).  That said, I'm totally open to other descriptors too -- would just wanna see a word that doesn't imply that we're somehow "changing" the inputs, you know?

Date: 2024-07-02

## Status

Under Review

## Context

> Issues were discovered in a few spots when accepting arbitrary keycloak client ids and setting them into the Store without sanitization. IDs like https://google.com/ failed during `Store.setItem` due to the `/` character and returned 422. 

Reviewer: ```suggestion
Issues were discovered in a few spots when accepting arbitrary keycloak client ids (i.e. URIs) as Store keys without sanitization. IDs like 'https://google.com/' fail during `Store.setItem()` calls due to '/' characters being interpreted as _separators within a JSON data path_ by the underlying implementation (resulting in a thrown error and returning a 422).
```
Just wanted to make it a bit more explicit for readers who weren't privy to our team convos.

## Decision

* Base64 encode the key after receiving it from the user and before setting it into the Store.
* Update the Store key prefix with v2: `/data/${capabilityName}-v2-${key}`

> * Migrate before calling `this.#onReady()` in the `#receive` function of `src/lib/controller/store.ts`, checking if each key that matches the old prefix, if so, migrating to the new prefix with base64 encoding. 

Reviewer: ```suggestion
* Migrate before calling `this.#onReady()` in the `#receive` function of `src/lib/controller/store.ts`: check each key to see if it matches the old prefix, and if so, migrating it to the new prefix with base64 encoding.
```
Extra words.

> * Add a new pepr command for viewing store items. like `npx pepr view-store`, maybe a `kubectl` plugin or alias too.

Reviewer: ```suggestion
* Add a new `pepr` command for viewing store items--something like `npx pepr view-store`, or maybe a `kubectl` plugin and/or alias too.
```
Minor syntax stuff.

Author: committed below

* Enhance existing unit testing to cover the new base64 encoding and migration logic. Add a battery of new fuzz and property-based tests.

## Consequences

1. Users cannot easily read the PeprStore CR
2. We need to migrate existing PeprStore CRs so that the `Store.getItem/Store.removeItem` will not fail due to the base64 encoding.

> 3. Will slightly slow down the store startup since we need to check and possibly migrate store before onReady.

Reviewer: ```suggestion
3. Will slightly slow down the store startup since we need to check and possibly migrate store content before onReady.
```

> 4. We need to build a new pepr command for viewing store items. like `npx pepr view-store`, maybe a `kubectl` plugin or alias too.

Reviewer: ```suggestion
4. We need to build a new `pepr` command for viewing store items. Something like `npx pepr view-store`, or maybe a `kubectl` plugin and/or alias too.
```
Minor syntax stuff.
